src/particles/particle_creation.hpp

```diff
@@ -135,6 +139,254 @@ template <ParticleType particleType> struct ParticleCreationTraits {
+						mass_of_star = std::pow(mass_of_star, 1. / (1. - alpha));
+						p.rdata(mass_idx) = mass_of_star;
+
+						total_momx += p.rdata(mass_idx + 1) * p.rdata(mass_idx);
+						total_momy += p.rdata(mass_idx + 2) * p.rdata(mass_idx);
+						total_momz += p.rdata(mass_idx + 3) * p.rdata(mass_idx);
+
+						p.idata(evolution_stage_index) = interpolate_fate(p.rdata(mass_idx));
+						p.rdata(birth_time_index + 1) = interpolate_death_time(p.rdata(mass_idx));
+						;
+					}
+				}
+
+				if (num_particles > 1) { // Update momentum of the low mass star if there is(are) high mass star(s) in the cell
+					const int p_idx = 0;
+					auto &plow = particles[p_idx]; // NOLINT
+					plow.rdata(mass_idx + 1) = -total_momx / plow.rdata(mass_idx);
+					plow.rdata(mass_idx + 2) = -total_momy / plow.rdata(mass_idx);
+					plow.rdata(mass_idx + 3) = -total_momz / plow.rdata(mass_idx);
+				}
+
+				const double factor = (cell_mass - particle_mass) / cell_volume / state_arr(i, j, k, HydroSystem<problem_t>::density_index);
+
+				// Update the cell density to reflect mass conversion into stars
+				state_arr(i, j, k, HydroSystem<problem_t>::density_index) *= factor;
+
+				// Update the cell momentum to make sure velocities don't change
+				state_arr(i, j, k, HydroSystem<problem_t>::x1Momentum_index) *= factor;
+				state_arr(i, j, k, HydroSystem<problem_t>::x2Momentum_index) *= factor;
+				state_arr(i, j, k, HydroSystem<problem_t>::x3Momentum_index) *= factor;
+
+				// Update internal energy to relect mass change
+				state_arr(i, j, k, HydroSystem<problem_t>::internalEnergy_index) *= factor;
+
+				// Update total energy
+				state_arr(i, j, k, HydroSystem<problem_t>::energy_index) *= factor;
+
+				// Update mass scalars
+				if (nscalars > 0) {
+					for (int nn = 0; nn < nscalars; ++nn) {
+						state_arr(i, j, k, HydroSystem<problem_t>::scalar0_index + nn) *= factor;
+					}
+				}
+			}
+		}
+	};
+
+	// Main method to create particles - uses the helper implementation
+	template <typename problem_t, typename ContainerType>
+	static void createParticles(ContainerType *container, int mass_idx, amrex::MultiFab &state, int lev, amrex::Real current_time, amrex::Real dt,
+				    int evolution_stage_index = -1, int birth_time_index = -1)
+	{
+		// Use the common implementation with our checker and creator types
+		ParticleCreationImpl::createParticlesImpl<problem_t, ContainerType,
+							  ParticleCreationTraits<ParticleType::StochasticStellarPop>::template ParticleChecker,
+							  ParticleCreationTraits<ParticleType::StochasticStellarPop>::template ParticleCreator>(
+		    container, mass_idx, state, lev, current_time, dt, evolution_stage_index, birth_time_index);
+	}
+}; // ParticleCreationTraits<ParticleType::StochasticStellarPop>
+
+double ParticleCreationTraits<ParticleType::StochasticStellarPop>::fstar_high;	    // NOLINT
+double ParticleCreationTraits<ParticleType::StochasticStellarPop>::m_star_high_avg; // NOLINT
```

Comment:
Global variables for this kind of data are fine, but they can't be in header files: if two different *.cpp files include this header, they will have two different versions of this variable implicitly defined. This problem is generically referred to as an "ODR violation." (This is an unfortunate "feature" of C++...)

These variables either need to be moved into a *.cpp file, or you need to add `inline` to these variables.

Reply:
The (very) technical details of why this is a problem are here: https://en.cppreference.com/w/cpp/language/definition